### DISCUSSION TITLE
Store files as value?

### DISCUSSION THREAD

**Discussion by pwFoo** [azu/kvs#40] [Q&A]
Is it possible to save files? localforage handle files and serialization of objects out of the box.

**Comment by azu** [comment]
Perhaps what you are looking for is a feature like [Snapshots](https://unstorage.unjs.io/utils#snapshots) in unstorage.
kvs does not prodivce special API, but it is possible to iterate over storage.
`storage` is [AsyncIterable object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/asyncIterator).

- https://stackblitz.com/edit/kvs-snapshot-restore?file=index.js

```js
// polyfill for Array.fromAsync
import arrayFromAsync from 'array-from-async';
import { kvsEnvStorage } from '@kvs/env';
import * as fs from 'node:fs/promises';
// open database and initialize it
const storage = await kvsEnvStorage({
  name: 'database-name',
  version: 1,
});

// set
{
  await storage.set('a1', 'string');
  await storage.set('b2', 42);
  await storage.set('c3', false);
  console.log('Current', await arrayFromAsync(storage));
} 
// SAVE snapshot
{
  // Note: Node.js does not Array.fromAsync yet
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/fromAsync
  const snapshot = await arrayFromAsync(storage);
  // save snapshot
  await fs.writeFile('./snapshot.json', JSON.stringify(snapshot), 'utf-8');
}
// =====
// clean
{
  await storage.clear();
  console.log('Cleared', await arrayFromAsync(storage));
}
// =====
// RESTORE snapshot
{
  const savedSnapshot = JSON.parse(
    await fs.readFile('./snapshot.json', 'utf-8')
  );
  for (const entry of savedSnapshot) {
    await storage.set(entry[0], entry[1]);
  }
  console.log('Restored', await arrayFromAsync(storage));
}
```

📝 [Array.fromAsync()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/fromAsync) is [Stage 3](https://github.com/tc39/proposals#stage-3) proposal. Node.js does not support it yet, but it will be implemented in near future.

**Reply by pwFoo** [reply]
Hi,
I looked into localforage (browser only) which supports just to save files / images. 

Searching a solution working in browser an node to save strings / serialized objects and files / images to storage.
```
localForage supports storing all native JS objects that can be serialized to JSON, as well as ArrayBuffers, Blobs, and TypedArrays. Check the API docs for a full list of types supported by localForage.
```

**Comment by pwFoo** [comment]
Hi @azu,
stumbled across that old discussion. Thanks for the hint to unstorage. It's a great project too...

I have a working method to convert blobs, images and files to base64 and back. So store files isn't a blocker anymore, but there is still a problem with my data structure...

My data is a object tree which I try to store in atomic parts property => valWithMetadataJsonString , but...

would be nice to filter by metadata like write timestamp
I'm not sure how to build my object tree based on stored flatten data...
First idea was key based on <objId>/<property> and query <objId>/.*... But kv-storgae isn't designed to get items by key with wildcard or regexp.

Any idea how to store and load a object tree in atomic parts with kvs?
